src/matcher/many_patterns/naive.rs

```diff
@@ -59,14 +59,12 @@ impl<C, I> Default for NaiveManyMatcher<C, I> {
     }
 }
 
-impl<K, P, D, I> PortMatcher<D> for NaiveManyMatcher<Constraint<K, P>, I>
+impl<P, D, I> PortMatcher<D> for NaiveManyMatcher<Constraint<indexing::Key<I, D>, P>, I>
```

Comment:
This is the kind of thing where I think, maybe my suggestions were not quite right. (Maybe! Obviously, that never happens ;-) ;-))

I wonder how plausible it is to rebase this PR to *not* do the "drastic generic simplification", then just merge it ;-) (it's looking good now I think), and try the generic simplification as a separate PR - we might want to iterate a little bit, and/or maybe do only some of the ideas.

That said, some other things in that commit (e.g. `type BindingResult = Result`) are not really generic simplifications and I'd be happy for that to go in straightaway

Reply:
I'm not sure why you are having second thoughts :) I find it overall simplified things quite a bit, so I'm happy with the improvement.

I'll improve if further ideas come to mind.